genfs.c

```diff
@@ -0,0 +1,324 @@
+    va_end(args);
+    exit(EXIT_FAILURE);
+}
+
+_Noreturn static void genfs_err_inode(errcode_t const err,
+                                      const char *const name,
+                                      const char *const msg,
+                                      ext2_ino_t const ino,
+                                      int const namelen,
+                                      const char *const dirent_name) {
+    genfs_err(err, "%s of inode %" PRIu32 " (name %s/%.*s)", msg, ino, name, namelen, dirent_name);
+}
+
+static errcode_t qubes_write_inode_full(ext2_filsys fs, ext2_ino_t ino, struct ext2_inode_large *inode) {
+    return ext2fs_write_inode_full(fs, ino, ext2fs_inode(inode), sizeof *inode);
+}
+
+static errcode_t qubes_read_inode_full(ext2_filsys fs, ext2_ino_t ino, struct ext2_inode_large *inode) {
+    return ext2fs_read_inode_full(fs, ino, ext2fs_inode(inode), sizeof *inode);
+}
+
+static bool mark_immutable;
+
+int main(int argc, char **argv) {
+    initialize_ext2_error_table();
+    if (argc == 4) {
+        if (!strcmp("immutable=yes", argv[3]))
+            mark_immutable = true;
+        else if (!strcmp("immutable=no", argv[3]))
+            mark_immutable = false;
+        else
+            goto misuse;
+    } else if (argc != 2) {
+misuse:
+        errx(1, "Usage: genfs FILESYSTEM [UNAME immutable=[yes|no]]");
+    }
+    const char *const filesystem = argv[1], *const uname = argv[2];
+    return real_main(filesystem, uname);
+}
+
+struct qubes_genfs_data {
+    ext2_filsys fs;
+    const char *const uname;
+};
+
+static int recursive_relabel(ext2_ino_t dir __attribute__((unused)),
+                             int entry __attribute__((unused)),
+                             struct ext2_dir_entry *dirent,
+                             int offset __attribute__((unused)),
+                             int blocksize __attribute__((unused)),
+                             char *buf __attribute__((unused)),
+                             void *priv_data);
+
+static int root_iterate_callback(ext2_ino_t dir __attribute__((unused)),
+                                 int entry __attribute__((unused)),
+                                 struct ext2_dir_entry *dirent,
+                                 int offset __attribute__((unused)),
+                                 int blocksize __attribute__((unused)),
+                                 char *buf __attribute__((unused)),
+                                 void *priv_data) {
+    struct qubes_genfs_data *data = priv_data;
+    int const name_len = ext2fs_dirent_name_len(dirent);
+    errcode_t err;
+    process_dirent(data->fs, "", dirent->inode, name_len, dirent->name,
+                   "system_u:object_r:modules_object_t:s0", mark_immutable);
+    if (!strncmp(dirent->name, "firmware", (size_t)name_len)) {
+        return 0;
+        struct qubes_genfs_data relabel_data = {
+            .fs = data->fs,
+            .uname = NULL,
```

Comment:
Are you sure you don't need to relabel firmware too? On a non-qubes Fedora it's `system_u:object_r:lib_t:s0`, not `system_u:object_r:modules_object_t:s0`